src/codegen.cpp

```diff
@@ -533,6 +533,51 @@ static AttributeList get_attrs_zext(LLVMContext &C)
                 {Attributes(C, {Attribute::ZExt})});
 }
 
+static AttributeList get_attrs_ipoeffects(LLVMContext &C, jl_code_instance_t *ci)
+{
+    uint32_t e = ci->ipo_purity_bits;
+    uint8_t consistent = (e >> 0) & 0x07;
+    uint8_t effect_free = (e >> 3) & 0x03;
+    uint8_t nothrow = (e >> 5) & 0x01;
+    uint8_t terminates = (e >> 6) & 0x01;
+    uint8_t notaskstate = (e >> 7) & 0x01;
+    uint8_t inaccessiblememonly = (e >> 8) & 0x03;
+    uint8_t nonoverlayed = (e >> 10) & 0x01;
+#if JL_LLVM_VERSION >= 140000
+    AttrBuilder attr(C);
+#else
+    AttrBuilder attr;
+#endif
+    if (consistent == 0){
+        attr.addAttribute("consistent");
+    }
+    if (effect_free == 0)
+        attr.addAttribute("effect_free");
+    if (nothrow == 1){
+        attr.addAttribute("nothrow");
+        attr.addAttribute(Attribute::NoUnwind); // Is this even correct, because nothrow allows for try/catch
```

Comment:
I'm not familiar with the finer points of effects analysis and if this holds true there too, but out of memory errors would violate nounwind, so functions that allocate cannot be annotated with this.

Reply:
Yeah I would love to get @aviatesk opinion into what is correct here, we maybe have to add another effect or something of the sort here. Because for something like `sin` to be able to be moved by LICM it needs `readnone` and `nounwind`.

Reply:
Also the safepoint that was added at function entries might break every assumption we want to make here 😆

Reply:
I just read LLVM's documentation on `nounwind` and it says:
```
This function attribute indicates that the function never raises an exception. If the function does raise an exception, its runtime behavior is undefined. However, functions marked nounwind may still trap or generate asynchronous exceptions. Exception handling schemes that are recognized by LLVM to handle asynchronous exceptions, such as SEH, will still provide their implementation defined semantics.
```
From this I think our definition of `:nothrow` can't be directly translated into `nounwind` since our `:nothrow semantics allow something like:
```julia
Base.@assume_effects :nothrow function raise_but_caught(a)
    try
        throw(a)
    catch a
        return a
    end
end
```

---

Having said that, this kind of case (where `:nothrow` is marked on a method that may `throw` something) is _very_ rare, since we usually taint the `:nothrow` effect bit immediately when we see any `throw` call and thus that case only happens when the user declares it explicitly.

We can set up another (non-exported) effect bit `:nothrow_strict` that does not allow any exceptions no matter they are caught or not, and also does not allow any user annotation.

Reply:
that still seems scary. we try pretty hard to avoid C style UB

Reply:
Yeah, having some more strict effect analysis for the LLVM case is probably needed here.

Reply:
I would be trivial to set up a new effect bit that represents the strict `:nothrow`. Do you want me to work on it?

Reply:
That would be awesome! If it's not too much of a hassle :)